pkg/registry/authorization/selfsubjectaccessreview/rest_patch.go

```diff
@@ -0,0 +1,49 @@
+package selfsubjectaccessreview
+
+import (
+	"strings"
+
+	"k8s.io/apiserver/pkg/authentication/user"
+
+	authorizationv1 "github.com/openshift/api/authorization/v1"
+	authorizationscope "github.com/openshift/apiserver-library-go/pkg/authorization/scope"
+)
+
+func userWithRequiredScopes(userToCheck user.Info) user.Info {
+	userExtra := userToCheck.GetExtra()
+	if userExtra == nil || !scopesNeedUserFull(userExtra[authorizationv1.ScopesKey]) {
+		return userToCheck
+	}
+
+	userExtraCopy := make(map[string][]string)
+	for k, v := range userExtra {
+		userExtraCopy[k] = v
+	}
+	userExtraCopy[authorizationv1.ScopesKey] = append(userExtraCopy[authorizationv1.ScopesKey], authorizationscope.UserFull)
+
+	userWithFullScope := &user.DefaultInfo{
+		Name:   userToCheck.GetName(),
+		UID:    userToCheck.GetUID(),
+		Groups: userToCheck.GetGroups(),
+		Extra:  userExtraCopy,
+	}
+
+	return userWithFullScope
+}
+
+// a self-SAR request must be authorized as if it has either the full user's permissions
+// or the permissions of the user's role set on the request (if applicable) in order
+// to be able to perform the access review
+func scopesNeedUserFull(scopes []string) bool {
```

Comment:
this should only return true if
1. the only scope is self-sar
2. the scope is self-sar + info
3. the scope is self-sar + info + list-project
4. the scope is self-sar + list-project

everything else should be return false.  otherwise we're accidentally escalating the permissions the token granted rights for.  The cases listed above as ok are cases where "this is a token about me to learn about my rights and honor them"

Reply:
@deads2k based on our discussion, I've updated `scopesNeedUserFull()` to specifically accept explicit scope sets. I've also extended the unit tests with basically all combinations of the current scopes.

One (minor?) caveat of the current implementation is that it'll sort the scopes slice in the user context in-place; I don't think this would be a problem, so I chose to avoid a slice copy. Let me know in case you think this could be an issue.